### PR TITLE
[spark-3.0-snapshot] Basic implementation for CREATE TABLE and ALTER TABLE.

### PR DESCRIPTION
**Pull request by jose-torres**
Basic implementation for CREATE TABLE and ALTER TABLE.

Note that because of some known interactions with Spark, not all the tests quite pass right now. All of the newly added ones to test the DDL functionality pass.

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -22,7 +22,7 @@ crossScalaVersions := Seq("2.12.8", "2.11.12")
 
 scalaVersion := crossScalaVersions.value.head
 
-resolvers += "Apache Spark snapshot repository" at "https://repository.apache.org/content/repositories/snapshots"
+resolvers += Resolver.mavenLocal
```
Comment (brkyvz) [inline build.sbt]: you should probably revert this change. The hope is that all fixes will be in Spark 3.0 some time soon

File: src/main/scala/io/delta/DeltaExtensions.scala [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.delta
+
+import org.apache.spark.sql.SparkSessionExtensions
+import org.apache.spark.sql.delta.analysis.DeltaAnalysis
+
+class DeltaExtensions extends Function1[SparkSessionExtensions, Unit] {
```
Comment (brkyvz) [inline src/main/scala/io/delta/DeltaExtensions.scala]: Can you add docs to this? @zsxwing will use this to inject SQL parser commands as well

File: src/main/scala/org/apache/spark/sql/delta/analysis/DeltaAnalysis.scala [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.delta.catalog.DeltaTableV2
+import org.apache.spark.sql.delta.sources.DeltaDataSource
+import org.apache.spark.sql.execution.datasources.LogicalRelation
+import org.apache.spark.sql.execution.{FilterExec, ProjectExec}
+import org.apache.spark.sql.execution.datasources.v2.DataSourceV2Relation
+import org.apache.spark.sql.types.{DataType, StructType}
+import org.apache.spark.sql.util.SchemaUtils
+
+// scalastyle:off
```
Comment (brkyvz) [inline src/main/scala/org/apache/spark/sql/delta/analysis/DeltaAnalysis.scala]: remove this

File: src/main/scala/org/apache/spark/sql/delta/catalog/DeltaCatalog.scala [hunk 4]
```diff
@@ -0,0 +1,343 @@
...
+import org.apache.spark.sql.sources.InsertableRelation
+import org.apache.spark.sql.sources.v2.{StagedTable, SupportsWrite, Table, TableCapability}
+import org.apache.spark.sql.sources.v2.TableCapability._
+import org.apache.spark.sql.sources.v2.writer.{V1WriteBuilder, WriteBuilder}
+import org.apache.spark.sql.types.{StructField, StructType}
+import org.apache.spark.sql.util.CaseInsensitiveStringMap
+
+import scala.util.control.NonFatal
+
+class DeltaCatalog(val spark: SparkSession) extends V2SessionCatalog(spark.sessionState)
```
Comment (brkyvz) [inline src/main/scala/org/apache/spark/sql/delta/catalog/DeltaCatalog.scala]: We should extend the DelegatingCatalogExtension now

File: src/main/scala/org/apache/spark/sql/delta/catalog/DeltaCatalog.scala [hunk 5]
```diff
@@ -0,0 +1,343 @@
...
+  // We keep this as an internal class because it needs to call into some catalog internals for the
+  // eventual table creation.
+  private class StagedDeltaTableV2(
+      ident: Identifier,
+      override val schema: StructType,
+      val partitions: Array[Transform],
+      override val properties: util.Map[String, String]) extends StagedTable with SupportsWrite {
+    override def name(): String = ident.name()
+
+    override def abortStagedChanges(): Unit = {}
```
Comment (brkyvz) [inline src/main/scala/org/apache/spark/sql/delta/catalog/DeltaCatalog.scala]: We can have abort clean up the _delta_log directory if we did write out something

File: src/main/scala/org/apache/spark/sql/delta/catalog/DeltaCatalog.scala [hunk 6]
```diff
@@ -0,0 +1,343 @@
...
+
+    override def commitStagedChanges(): Unit = {}
```
Comment (brkyvz) [inline src/main/scala/org/apache/spark/sql/delta/catalog/DeltaCatalog.scala]: ideally, this is when the table creation should happen

File: src/main/scala/org/apache/spark/sql/delta/catalog/DeltaCatalog.scala [hunk 7]
```diff
@@ -0,0 +1,343 @@
...
+      }
+      Some(oldTable)
+    } else {
+      None
+    }
+  }
+
+  // We keep this as an internal class because it needs to call into some catalog internals for the
+  // eventual table creation.
+  private class StagedDeltaTableV2(
```
Comment (brkyvz) [inline src/main/scala/org/apache/spark/sql/delta/catalog/DeltaCatalog.scala]: I would change the logic in this class a little bit if possible, but we can do that in a follow up.

File: src/main/scala/org/apache/spark/sql/delta/catalog/DeltaTableV2.scala [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+  private var forceOverwrite = false
+
+  override def truncate(): WriteIntoDeltaBuilder = {
+    forceOverwrite = true
+    this
+  }
+
+  override def buildForV1Write(): InsertableRelation = {
+    new InsertableRelation {
+      override def insert(data: DataFrame, overwrite: Boolean): Unit = {
```
Comment (brkyvz) [inline src/main/scala/org/apache/spark/sql/delta/catalog/DeltaTableV2.scala]: overwrite will never be true when using the V1 Write fallback API.

File: src/main/scala/org/apache/spark/sql/delta/catalog/DeltaTableV2.scala [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+            if (nameParts.size != 1) {
+              throw new IllegalArgumentException(s"Invalid nested partition column $nameParts")
+            }
+            nameParts.head
+          case transform =>
+            throw new IllegalArgumentException(s"Unsupported partition transform $transform")
+        }
+        WriteIntoDelta(
+          log,
+          // TODO: other save modes?
```
Comment (brkyvz) [inline src/main/scala/org/apache/spark/sql/delta/catalog/DeltaTableV2.scala]: We only have SaveMode.Overwrite if `truncate()` or `overwrite(filters: Array[Filter])` is called. Otherwise it's always Append mode in DataSourceV2.

File: src/main/scala/org/apache/spark/sql/delta/commands/CreateDeltaTableCommand.scala [hunk 10]
```diff
@@ -0,0 +1,270 @@
...
+      sparkSession.sessionState.catalog.createTable(tableWithoutSchema, ignoreIfExists = true,
+        validateLocation = false)
+
+      Nil
+    }
+  }
+
+
+  private def assertPathEmpty(
+                               sparkSession: SparkSession,
```
Comment (brkyvz) [inline src/main/scala/org/apache/spark/sql/delta/commands/CreateDeltaTableCommand.scala]: nit: indentation

File: src/main/scala/org/apache/spark/sql/delta/sources/DeltaDataSource.scala [hunk 11]
```diff
@@ -43,14 +45,28 @@ class DeltaDataSource
...
     // TODO Remove this when upgrading to Spark 3.0.0
     spark.conf.set("spark.sql.legacy.sources.write.passPartitionByAsOptions", "true")
   }
 
+  override def getTable(options: CaseInsensitiveStringMap): Table = {
+    val path = options.get("path")
+    DeltaTableV2(DeltaLog.forTable(SparkSession.active, path))
+  }
+
+  override def getTable(
```
Comment (brkyvz) [inline src/main/scala/org/apache/spark/sql/delta/sources/DeltaDataSource.scala]: this doesn't exist yet, so we should remove it for now

File: src/test/scala/org/apache/spark/sql/delta/DeltaDDLSuite.scala [hunk 12]
```diff
@@ -0,0 +1,981 @@
...
+      .set("spark.sql.extensions", classOf[DeltaExtensions].getName)
+  }
+
+  test("create table with schema and path") {
+    withTempDir { dir =>
+      withTable("delta_test") {
+        sql(s"""
+               |CREATE TABLE delta_test(a LONG, b String)
+               |USING delta
+               |OPTIONS('path'='${dir.getCanonicalPath}')""".stripMargin)
```
Comment (brkyvz) [inline src/test/scala/org/apache/spark/sql/delta/DeltaDDLSuite.scala]: can we change this to (or also test with) `LOCATION '${dir.getCanonicalPath}'`

Comment (brkyvz) [reply src/test/scala/org/apache/spark/sql/delta/DeltaDDLSuite.scala]: I guess there are tests below, so this is fine

File: src/test/scala/org/apache/spark/sql/delta/DeltaDDLSuite.scala [hunk 13]
```diff
@@ -0,0 +1,981 @@
...
+          (1, "IT", "Alice"),
+          (2, "CS", "Bob"),
+          (3, "IT", "Carol")).toDF("id", "dept", "name")
+        df.write.format("delta").partitionBy("name", "dept").save(path)
+
+        sql(s"CREATE TABLE delta_test USING delta LOCATION '$path'")
+
+        def checkDescribe(describe: String): Unit = {
+          // TODO: The order of these columns matches their definition in the DF and not the order
+          // specified in partitionBy, whereas DBR matches the partition by. Is that okay?
```
Comment (brkyvz) [inline src/test/scala/org/apache/spark/sql/delta/DeltaDDLSuite.scala]: should be fine

**Comment by jose-torres** [comment]
Addressed comments, except as suggested deferring StagedDeltaTableV2 changes to a followup PR. I agree there's probably a better way to structure this class.

File: src/test/scala/org/apache/spark/sql/delta/DeltaDDLSuite.scala [hunk 14]
```diff
@@ -0,0 +1,981 @@
...
+    withTempPath { path =>
+      withTable("src", "t1", "t2", "t3", "t4", "t5", "t6") {
+        sql("CREATE TABLE src(i int, p string) USING delta PARTITIONED BY (p) " +
+          "TBLPROPERTIES('delta.randomizeFilePrefixes' = 'true') " +
+          s"LOCATION '${path.getAbsolutePath}'")
+        sql("INSERT INTO src SELECT 1, 'a'")
+
+        // CREATE TABLE without specifying anything works
+        sql(s"CREATE TABLE t1 USING delta LOCATION '${path.getAbsolutePath}'")
+        checkAnswer(spark.table("t1"), Row(1, "a"))
```
Comment (brkyvz) [inline src/test/scala/org/apache/spark/sql/delta/DeltaDDLSuite.scala]: this isn't passing tests

File: src/test/scala/org/apache/spark/sql/delta/DeltaDDLSuite.scala [hunk 15]
```diff
@@ -0,0 +1,981 @@
...
+      // External catalog does not contain the schema and partition column names.
+      assert(table.schema == new StructType())
+      assert(table.partitionColumnNames == Seq("a"))
+
+      sql("INSERT INTO delta_test SELECT 1, 'a'")
+
+      val path = new File(new File(table.storage.locationUri.get), "a=1")
+      assert(path.listFiles().nonEmpty)
+
+      checkDatasetUnorderly(
```
Comment (brkyvz) [inline src/test/scala/org/apache/spark/sql/delta/DeltaDDLSuite.scala]: this isn't passing tests.

File: src/test/scala/org/apache/spark/sql/delta/DeltaAlterTableSuite.scala [hunk 16]
```diff
@@ -0,0 +1,486 @@
...
+        |ALTER TABLE delta_test
+        |SET TBLPROPERTIES (
+        |  'delta.logRetentionDuration' = '2 weeks',
+        |  'delta.checkpointInterval' = '20',
+        |  'key' = 'value'
+        |)""".stripMargin)
+
+      val deltaLog = DeltaLog.forTable(spark, TableIdentifier("delta_test"))
+      val snapshot1 = deltaLog.update()
+      assert(snapshot1.metadata.configuration == Map(
```
Comment (brkyvz) [inline src/test/scala/org/apache/spark/sql/delta/DeltaAlterTableSuite.scala]: This isn't passing tests

File: src/test/scala/org/apache/spark/sql/delta/DeltaAlterTableSuite.scala [hunk 17]
```diff
@@ -0,0 +1,486 @@
...
+        |ALTER TABLE delta_test
+        |SET TBLPROPERTIES (
+        |  'dEltA.lOgrEteNtiOndURaTion' = '1 weeks',
+        |  'DelTa.ChEckPoiNtinTervAl' = '5',
+        |  'key' = 'value1'
+        |)""".stripMargin)
+
+      val deltaLog = DeltaLog.forTable(spark, TableIdentifier("delta_test"))
+      val snapshot1 = deltaLog.update()
+      assert(snapshot1.metadata.configuration == Map(
```
Comment (brkyvz) [inline src/test/scala/org/apache/spark/sql/delta/DeltaAlterTableSuite.scala]: this isn't passing tests

File: src/test/scala/org/apache/spark/sql/delta/DeltaAlterTableSuite.scala [hunk 18]
```diff
@@ -0,0 +1,486 @@
...
+      assert(ex1.getCause.getMessage.contains("randomPrefixLength needs to be greater than 0."))
+
+      val ex2 = intercept[SparkException] {
+        sql("ALTER TABLE delta_test SET TBLPROPERTIES ('delta.randomPrefixLength' = 'value')")
+      }
+      assert(ex2.getCause.getMessage.contains("randomPrefixLength needs to be greater than 0."))
+    }
+  }
+
+  test("ADD COLUMNS - simple") {
```
Comment (brkyvz) [inline src/test/scala/org/apache/spark/sql/delta/DeltaAlterTableSuite.scala]: this isn't passing tests

File: src/test/scala/org/apache/spark/sql/delta/DeltaAlterTableSuite.scala [hunk 19]
```diff
@@ -0,0 +1,486 @@
...
+import org.apache.spark.sql.delta.DeltaConfigs.CHECKPOINT_INTERVAL
+import org.apache.spark.sql.delta.catalog.DeltaCatalog
+import org.apache.spark.sql.delta.sources.DeltaSQLConf
+import org.apache.spark.sql.functions._
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.test.SharedSparkSession
+import org.apache.spark.sql.types.{ArrayType, MapType, StructType}
+import org.apache.spark.util.Utils
+
+class DeltaAlterTableSuite extends QueryTest with SharedSparkSession {
```
Comment (brkyvz) [inline src/test/scala/org/apache/spark/sql/delta/DeltaAlterTableSuite.scala]: most of the tests here are not passing. Something seems to be broken

Comment (brkyvz) [reply src/test/scala/org/apache/spark/sql/delta/DeltaAlterTableSuite.scala]: I think it's because you haven't overloaded `loadTable`

**Comment by brkyvz** [comment]
Only a 115 failed tests! Not bad. Those will be fixed in OSS hopefully soon. I'm merging this large PR so that we can start adding more stuff.
